File: Golang/Easy/3354_Make_Array_Elements_Equal_to_Zero/solution.go
package makearrayelementsequaltozero

func countValidSelections(nums []int) int {
	count := 0
	for i, _ := range nums {
		if nums[i] != 0 {
			continue
		}

		right := append([]int{}, nums...)
		right = simulateMovement(i, right, 1)
		if isAllZero(right) {
			count += 1
		}

		left := append([]int{}, nums...)
		left = simulateMovement(i, left, -1)
		if isAllZero(left) {
			count += 1
		}
	}
	return count
}

func simulateMovement(p int, array []int, sign int) []int {
	for p < len(array) && p >= 0 {
		if array[p] == 0 {
			p += sign
			continue
		} else {
			array[p] -= 1
		}

		sign *= -1
		p += sign
	}
	return array
}

func isAllZero(array []int) bool {
	for _, num := range array {
		if num != 0 {
			return false
		}
	}
	return true
}

File: Python/Easy/3354_Make_Array_Elements_Equal_to_Zero/solution.py
class Solution:
    def countValidSelections(self, nums: list[int]) -> int:
        count = 0
        for i in range(len(nums)):
            if nums[i] != 0:
                continue
            right = Solution.simulateMovement(i, nums.copy(), 1)
            if Solution.isAllZero(right):
                count += 1

            left = Solution.simulateMovement(i, nums.copy(), -1)
            if Solution.isAllZero(left):
                count += 1
        return count

    def simulateMovement(p, array, sign):
        while p < len(array) and p >= 0:
            if array[p] == 0:
                p += sign
                continue
            else:
                array[p] -= 1
            sign *= -1
            p += sign
        return array

    def isAllZero(nums):
        for i in range(len(nums)):
            if nums[i] != 0:
                return False
        return True